content/post/mnist-example.md

```diff
@@ -0,0 +1,155 @@
+
+```python
+@wallaroo.decoder(header_length=4, length_fmt=">I")
+def decode(bs):
+    return pickle.loads(bs)
+```
+
+### PCA Computation
+
+The first computation applies the PCA that has been previously loaded.
+
+```python
+@wallaroo.computation(name="PCA")
+def pca_transform(x):
+    return pca.transform([x])
+```
+
+### Logistic Regression computation
+
+The second computation applies the logistic regression, also previously loaded.
+
+```python
+@wallaroo.computation(name="Logistic Regression")
+def logistic_classify(x):
+    return logistic.predict(x)
+```
+
+### Encoder
+
+The encoder packs the result into a string and sends it across the wire with a framed header containing the length of the string.
+
+```python
+@wallaroo.encoder
+def encode(data):
+    s = str(data)
+    return struct.pack('>I{}s'.format(len(s)), len(s), s)
+```
+
+## Sending data to Wallaroo
+
+In order to send data into Wallaroo, we must use a special sender that knows how to pickle the images in the right way. We can create this sender by formatting our messages such that they match the working of the decoder:
+
+* 4 bytes representing the length of the message, followed by
+* a UTF-8 encoded string, containing the pickled object
+
+Our sender will send the entire MNIST dataset.
+
+```python
+import sys
+import socket
+import struct
+from sklearn import linear_model, decomposition, datasets
+
+import cPickle as pickle
+
+def send_message(conn, x):
+    msg = pickle.dumps(x)
+    conn.sendall(struct.pack('>I', len(msg)))
+    conn.sendall(msg)
+
+if __name__ == '__main__':
+    add = sys.argv[1].split(':')
+    wallaroo_input_address = (add[0], int(add[1]))
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    print 'connecting to Wallaroo on %s:%s' % wallaroo_input_address
+    sock.connect(wallaroo_input_address)
+    digits = datasets.load_digits()
+    for x in digits.data:
+        send_message(sock, x)
+```
+
+## Running our application
+
+To run our application, we need to follow these steps:
+
+- start a listener so we can view the output : `nc -l 7002`
+- start the Wallaroo application from within its directory: `PYTHONPATH=:.:../../../machida/ ../../../machida/build/machida --application-module digits --in 127.0.0.1:8002 --out 127.0.0.1:7002 --metrics 127.0.0.1:5001 --control 127.0.0.1:6000 --data 127.0.0.1:6001 --worker-name worker1 --external 127.0.0.1:5050 --cluster-initializer --ponythreads=1`
+- send our files to Wallaroo via our sender: `python sender.py`
+
+## Next steps
+
+There are obvious limitations to this basic example. For instance, there is no partitioning. And we of course realize that MNIST isn't a useful dataset beyond examples. A lot of extra functionality can be added to production-level code, but for the purpose of illustrating how to run scikit-learn algorithms in Wallaroo, we preferred to narrow the focus and reduce distractions.
```

Comment:
I'm not sure the limitations are "obvious" nor that people will "of course realize" that MNIST isn't useful beyond examples. That would depend on peoples knowledge of Wallaroo, ML, and/or the MNIST dataset. It would be great if we can explain why MNIST isn't useful beyond examples (one or two sentences would be great). Similarly, we can point out that there are additional wallaroo features not used (such as partitioning, etc) and link to them for further reading by folks.